src/librustc_lint/types.rs

```diff
@@ -427,42 +428,81 @@ enum FfiResult<'tcx> {
     },
 }
 
+fn is_zst<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, did: DefId, ty: Ty<'tcx>) -> bool {
+    tcx.layout_of(tcx.param_env(did).and(ty)).map(|layout| layout.is_zst()).unwrap_or(false)
+}
+
+fn ty_is_known_nonnull<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, ty: Ty<'tcx>) -> bool {
+    match ty.sty {
+        ty::FnPtr(_) => true,
+        ty::Ref(..) => true,
+        ty::Adt(field_def, substs) if field_def.repr.transparent() => {
+            if let AdtKind::Union = field_def.adt_kind() {
+                return false;
+            }
+
+            for field in &field_def.non_enum_variant().fields {
+                let field_ty = tcx.normalize_erasing_regions(
+                    ParamEnv::reveal_all(),
```

Comment:
> Is there a specific reason why it's used here?

Because I don't know what I'm doing 😛. I used `reveal_all()` because 3 out of 4 places that involved getting the field type used it:

- Doesn't use `reveal_all`: [1](https://github.com/rust-lang/rust/blob/6cc24f26036b28fb3366de86efe3da6c4464057a/src/librustc_lint/types.rs#L454).
- Uses `reveal_all`: [1](https://github.com/rust-lang/rust/blob/6cc24f26036b28fb3366de86efe3da6c4464057a/src/librustc_lint/types.rs#L514-L517), [2](https://github.com/rust-lang/rust/blob/6cc24f26036b28fb3366de86efe3da6c4464057a/src/librustc_lint/types.rs#L562-L565), [3](https://github.com/rust-lang/rust/blob/6cc24f26036b28fb3366de86efe3da6c4464057a/src/librustc_lint/types.rs#L514-L517).

If I shouldn't be doing this (or if I should be doing it only once but somewhere else, like in `is_repr_nullable_ptr`), please let me know!

Reply:
I am not quite sure either. It makes intuitive sense to me that *normalization* is needed in all of those places (including the one which currently isn't normalizing). But after some digging I am pretty sure that the use of reveal_all specifically in the normalization leads to undesirable behavior. 

For example, in [this program](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=76fdb36dcf516354a400647c26b64e87) using existential types, it makes the lint pierce through otherwise-opaque existentials and both
- accept code that is only FFI-safe because of the particular type underlying the existential, which users of that type shouldn't know or rely on
- leak the underlying type to the user in error messages when (correctly) rejecting uses of an existential type as FFI-unsafe

I believe if we didn't use reveal_all in those places (but just normalized with a `UserFacing` reveal mode) the lint would complain about both of those functions in terms of the existential types. IMO that is clearly better behavior, but this is also a pre-existing issue, so I'm not sure how to best proceed.